### DISCUSSION TITLE
Simplifying returns on non-void methods by caching return value

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#6120] [Language Ideas]
## Summary

At a very specific scenario, I came up with the idea of having the ability to store the return value and avoid specifying the expression to return when returning from the method. In other words, temporarily store the value that is to be returned upon returning from the method, and upon doing `return;`, the stored value is returned.

## Examples

Consider the example below:
```csharp
int a = CalculateA();
if (a > 1)
    return value;

int b = CalculateB();
if (b > 3)
    return value;

int c = CalculateC();
if (c > 4)
    return value;
```

With this proposal, it could be changed to:

```csharp
return = value;
int a = CalculateA();
if (a > 1)
    return;

int b = CalculateB();
if (b > 3)
    return;

int c = CalculateC();
if (c > 4)
    return;
```

Syntax is not final, and probably better alternatives exist that will better convey the feature's objective.

Using this feature does not prevent the user from doing `return x;`, which would simply return `x`, ignoring the locally stored return value.

## Motivation and Goals

When returning the same expression from multiple code paths, it doesn't feel obvious. In scenarios with few different returned expressions, this would remove the need for an explicitly created temporary variable to store the returned value, and also feel easier to track, as it will be more special to the eye.

## Drawbacks

This feature is very niche, and is better avoided in most places. The cost of implementing the feature far outweights its benefits and applications' impact. Non-confusing syntax is also difficult to come up with. It is also unclear if reading the current contents of the cached return value is meaningful to support, or even available.

**Comment by HaloFour** [comment]
I find the "before" code much clearer.  Assigning the return value to some magic local was how returning values from functions worked in VB Classic and it was always very error prone, I would not want to see that brought to C#.

**Comment by thinker227** [comment]
What would be required to tell the compiler to use this new syntax over the traditional syntax? Would the return value be required to be set at the start of the method?

Would this synthesized variable be write-only or could it be also read from? In this case, would it have the same restrictions as `out` parameters which follow essentially the same principle?

Would you be able to use `return x;` in methods in which `return` is treated as a variable?

Since this would literally just introduce a (potentially) write-only implict variable which acts as the return value of the method, are the benefits significant enough to justify this over just returning a regular variable?

And lastly, since you say "[t]his feature is very niche, and is better avoided in most places", then why implement this feature in the first place?

**Reply by Rekkonnect** [reply]
> What would be required to tell the compiler to use this new syntax over the traditional syntax? Would the return value be required to be set at the start of the method?

Nothing, the `return` local is always there, and simply removed from the resulting IL if never used (no assignment of the form `return = ...;`). Additionally, it has to be first assigned before used in `return;` statements.

> Would this synthesized variable be write-only or could it be also read from? In this case, would it have the same restrictions as `out` parameters which follow essentially the same principle?

It feels wrong to restrict to write-only, for there could be scenarios relying on the returning value. That being said, ` = return` wouldn't exactly feel great, but that's a syntax issue to resolve.

> Would you be able to use `return x;` in methods in which `return` is treated as a variable?

Yes, although not mentioned, nothing prevents you from returning a different expression.

> Since this would literally just introduce a (potentially) write-only implict variable which acts as the return value of the method, are the benefits significant enough to justify this over just returning a regular variable?

> And lastly, since you say "[t]his feature is very niche, and is better avoided in most places", then why implement this feature in the first place?

Already mentioned in the drawbacks, this is not very strong a feature.

**Comment by 333fred** [comment]
Largely a duplicate of https://github.com/dotnet/csharplang/discussions/142.
